src/download.c

```diff
@@ -452,6 +452,9 @@ static int perform_curl_io_and_complete(int count, bool bounded)
 	return 0;
 }
 
+/* Fixed window added to hysteresis upper bound before it is enforced. */
+#define BUFFER 5
```

Comment:
There is no chance for an upper bound bypass to result in running out of fds; this is purpose behind the conditional check below. The code comment above that check describes the reason for adding a bit extra, though "5" is an arbitrary number, it could be larger or smaller.